File: backend/forms/tax_year.py
"""Tax year utilities — shared logic for year detection and availability."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any


def active_tax_year(today: date | None = None) -> int:
    """Return the active tax year based on today's date.

    Rule: if today is any date in year Y, the active tax year is Y - 1.
    Example: any date in 2026 → returns 2025.
    """
    if today is None:
        today = date.today()
    return today.year - 1


def available_tax_years(sync_rows: list[dict[str, Any]]) -> list[int]:
    """Return years that have forms marked available in the DB.

    *sync_rows* is a list of dicts from ``forms_sync_status`` with at
    least ``tax_year`` and ``status`` keys.
    """
    years: set[int] = set()
    for row in sync_rows:
        if row.get("status") == "available":
            years.add(int(row["tax_year"]))
    return sorted(years)


def is_new_year_unlocked(
    year: int, sync_rows: list[dict[str, Any]]
) -> bool:
    """A year is unlocked if the DB has at least one row for that year
    with ``status='available'``.
    """
    for row in sync_rows:
        if int(row.get("tax_year", 0)) == year and row.get("status") == "available":
            return True
    return False

File: lib/taxYear.ts
/**
 * Tax year utilities — shared logic for year detection and availability.
 * Mirrors backend/forms/tax_year.py exactly.
 */

/**
 * Return the active tax year based on today's date.
 * Rule: if today is any date in year Y, the active tax year is Y - 1.
 * Example: any date in 2026 → returns 2025.
 */
export function activeTaxYear(today?: Date): number {
  const d = today ?? new Date();
  return d.getFullYear() - 1;
}

interface SyncRow {
  tax_year: number;
  status: string;
}

/**
 * Return years that have forms marked "available" in the DB.
 */
export function availableTaxYears(syncRows: SyncRow[]): number[] {
  const years = new Set<number>();
  for (const row of syncRows) {
    if (row.status === "available") {
      years.add(row.tax_year);
    }
  }
  return Array.from(years).sort((a, b) => a - b);
}

/**
 * A year is unlocked if the DB has at least one row for that year
 * with status="available".
 */
export function isNewYearUnlocked(year: number, syncRows: SyncRow[]): boolean {
  return syncRows.some(
    (row) => row.tax_year === year && row.status === "available"
  );
}
